pyrolite/mineral/normative.py

```diff
@@ -137,95 +133,233 @@ def endmember_decompose(
     modal = renormalise(modal)
     return modal
 
-
-def CIPW_norm(df, form="weight"):
+def CIPW_norm(data, form="weight"):
     """
     Standardised calcuation of estimated mineralogy from bulk rock chemistry.
-    Takes a dataframe of chemistry and creates a dataframe of estimated mineralogy.
-
+    Takes a dataframe of chemistry & creates a dataframe of estimated mineralogy.
+    
+    This is the CIPW norm of Verma et al. (2003)
+    
+    This version only uses major elements
     Parameters
     -----------
-    df : :class:`pandas.DataFrame`
+    df : :class:`p&as.DataFrame`
         Dataframe containing compositions to transform.
-
     Returns
     --------
     :class:`pandas.DataFrame`
-
     Notes
     -----
     This function is currently a stub.
     """
-
-    # devolatilise and renormalise
-
-    # deal with Fe oxidation
-
-    # convert to molar data
-
-    # aggregate like-elements to their most commom member
-    # Ferric : FeO, MnO, NiO, CoO
-    # Calcic : CaO, BaO, SrO
-    # Potassic : K2O, Rb2O, Cs2O
-    # Sodic : Na2O, Li2O
-    # Chromic : Cr2O3, V2O3
-
-    # recalculate 'molecular weigths' for each of these components, and the minerals
-
-    # Zircon
-
+    columns = ['SiO2','TiO2','Al2O3','Fe2O3','FeO','MnO','MgO','CaO','Na2O','K2O','P2O5','CO2','SO3']
```

Comment:
How's the best way to do this? Try to read the columns and then arrange them into the correct order, and then maybe impute missing columns as 0. Could also raise an error if the total is less than a certain threshold.

Reply:
I'd go with something along the lines of:
1. Check which columns are present in the given dataframe; add a `logger.warning(...)` message if anything critical is missing.
2. Use a `df.reindex(columns=<expected_columns>).fillna(0)`, and add a `logger.debug(...)` message noting what's been filled/added.
3. We'll need some kind of switch depending on what iron compositions are specified; if only one is given then some assumption might be needed to normalise it (potentially add a `logger.debug(..)` message for this too..), otherwise if FeO and Fe2O3 are present then perhaps its simpler as less assumptions are needed?